### ISSUE TITLE
FtpExists.Skip option will still overwrite files

### ISSUE THREAD

**Issue by dbeavon** [robinrodricks/FluentFTP#364]
**FTP OS:** Unix HP-UX

**FTP Server:** wuftpd-2.6.1 

**Computer OS:** Windows 10

I would like to upload without the risk of clobbering an existing file.  I use the UploadFile method with option FtpExists.Skip.

Despite specifying that I want to skip existing files, my upload operation will always overwrite the file that is there.
 
I suspect that the problem might be related to NLST.  It seems that when the NLST command is sent, the ftp server claims there are no files. See log below.

**Logs :**

```

# OpenActiveDataStream(PORT, "NLST /home/cool", 0)
Command:  PORT 172,30,11,212,241,91
Response: 200 PORT command successful.
Command:  NLST /home/cool
Response: 150 Opening BINARY mode data connection for file list.
Status:   Disposing FtpSocketStream...


```

I can use the GetListing method (LIST) as a workaround to search for the pre-existing files manually.  That isn't very ideal and it would be more comforting if the UploadFile method would check for the file before overwriting the existing one.

**Comment by robinrodricks** [comment]
This is definitely a bug. We'll have to find a working way to check for existing files on WuFTPD server.

**Comment by robinrodricks** [comment]
Hi,

I added detection of FTP Server software. Could you connect to your server and see if ServerType is being populated with WuFtp as it should?

**Comment by dbeavon** [comment]
Yes, I see the following:

>		ServerOS	Unix	FluentFTP.FtpOperatingSystem
>		ServerType	WuFTPd	FluentFTP.FtpServer
>		SystemType	"UNIX Type: L8"	string


Previously (version="19.2.2" ) I only had the system type (SystemType	"UNIX Type: L8")

Note that the overwrite behavior hasn't changed.  It still overwrites when I ask it to skip.

**Comment by robinrodricks** [comment]
Cool, but no idea how UploadFile calls name listing. Only UploadFiles does.
Which API are you using exactly?

**Comment by dbeavon** [comment]
I use the following call:

> 
>                // Upload the file if possible
>                var ResultFromUploadFile =  ClientObj.UploadFile(
>                    localPath: p_FullLocalFilePath,
>                    remotePath: RemotePathAndFile,
>                    existsMode: p_ExistsMode,
>                    createRemoteDir: false,
>                    verifyOptions: FtpVerify.None);


That, in turn, calls ...

- UploadFileFromFile
- which eventually calls UploadFileInternal
- which relies on a method that seems to be *misbehaving* : FileExists (FtpClient_Management)

The FileExists() is bypassing logic based on "Capabilities" (FluentFTP.FtpCapability is NONE):  It *doesn't* check file existence based on:
HasFeature(FtpCapability.SIZE)

... and it *doesn't* check file existence based on:
HasFeature(FtpCapability.MDTM))

(Those seem like the preferred methods of checking for a file, and the docs seem to say that they are supported, so I'm not sure why my capabilities are reported as NONE.  Here is a link..
https://www.wfms.org/wu-ftpd/)


In the end I reach the last part of the method, FileExists() that checks the file name (GetFtpFileName) to the items in a listing (GetNameListing)  ... and this *doesn't* behave correctly because the output from GetFtpFileName gives us only the name of the file, whatever123.txt ... while the listing provides the full paths (/usr/tmp/whatever123.txt) ... so the name is not going to be an exact match for the items in the list.

Hope this is clear.  Please let me know if not.

**Comment by robinrodricks** [comment]
That's awesomely clear and well explained, thank you.. but I would like to
see some of the logs pertaining to the start of the connection, where we
ask the server for it's capabilities.

Could you paste the log with the FEAT command?

**Comment by dbeavon** [comment]
Hi, I looked at my logs and FEAT is failing.  Good call.  

> Command:  SYST
> Response: 215 UNIX Type: L8
> Status:   Auto-detected UNIX listing parser
> Status:   Connecting to xxxxx
> Response: 220 ufpdev.ufpi.com FTP server (Revision 6.0 Version wuftpd-2.6.1 Fri Apr  1 07:44:09 GMT > 2011) ready.
> Status:   Detected FTP server: WuFTPd
> Command:  USER xxxxxxxxxx
> Response: 331 Password required for xxxxxxx.
> Command:  PASS ***
> Response: 230 User xxxxx logged in.
> Command:  FEAT
> Response: 500 'FEAT': command not understood.

I probably missed a FAQ, but what is the best way to manually assign the capabilities so that they include SIZE and MDTM (in spite of detecting that via FEAT).  Should I derive from FtpClient and assign the protected Capabilities property directly?


Also, sorry for the link to www.wfms.org above.  it is misleading because it references a version of wu-ftpd not found in HP-UX.  All I am working with is the HP-UX version of wu-ftpd 2.6.1. The following man page is a better match for the features I have in my local man pages.
http://nixdoc.net/man-pages/HP-UX/ftpd.1m.html

I tested them and both SIZE and MDTM are supported.  But oddly enough,  FEAT is not supported.

I appreciate all your help.  Truthfully you may not want to spend much time on the ideosyncracies of HP-UX's wu-ftpd.  The whole OS has a roadmap that ends after 2025 according to HPE.

Please let me know if there is a good way to ensure that the FileExists() returns true.  This github code is fairly easy to understand, and I would probably derive a new class and override the GetFeatures method to assign features manually.

**Comment by dbeavon** [comment]
I ended up overriding the Connect method in my derived class, and manually adjusting the "Capabilities" after the base version of the connect had completed.

It is too bad that FEAT isn't in this version of wu-ftpd ... I'm not sure how else you would know that about the ability to use SIZE and MDTM.

**Comment by robinrodricks** [comment]
Oh never mind, I'll fix it anyway. One fix leads to another, I always find. As of now I'll assume these capabilities when we connect to WuFtpd:

           ABOR           Abort previous command
           ACCT           Specify account (ignored)
           ALLO           Allocate storage (vacuously)
           APPE           Append to a file
           CDUP           Change to parent of current working directory
           CWD            Change working directory
           DELE           Delete a file
           EPSV           Sets the server to listen on a data port and wait
                          for a connection
           EPRT           Use extended address for data connection
           HELP           Give help information
           LIST           Give list files in a directory (ls -l)
           LPRT           Use long address for data connection
           LPSV           Sets the server to listen on a data port and wait
                          for a connection
           MKD            Make a directory
           MDTM           Show last modification time of file
           MODE           Specify data transfer mode
           NLST           Give name list of files in directory
           NOOP           Do nothing
           PASS           Specify password
           PASV           Prepare for server-to-server transfer
           PORT           Specify data connection port
           PWD            Print the current working directory
           QUIT           Terminate session
           REST           Restart incomplete transfer
           RETR           Retrieve a file
           RMD            Remove a directory
           RNFR           Specify rename-from file name
           RNTO           Specify rename-to file name
           SITE           Non-standard commands (see next section)
           SIZE           Return size of file
           STAT           Return status of server
           STOR           Store a file
           STOU           Store a file with a unique name
           STRU           Specify data transfer structure
           SYST           Show operating system type of server system
           TYPE           Specify data transfer type

Also, I must fix the bug for FileExists when it uses name listing.

**Comment by robinrodricks** [comment]
Can you check if this private build solves your problems? You don't need to use your derived class.. All fixes are automatically done, no additional config is required.

Just replace these DLLs into your nuget package dir:

[Release.zip](https://github.com/robinrodricks/FluentFTP/files/2913601/Release.zip)

**Comment by dbeavon** [comment]
Yes, I no longer have problems in the FileExists() method.

Oddly my capabilities in the watch window, while stepping into FileExists() method, are not what you had described.  They are only the following.  At first I thought my own code was still in effect.

>
> this.Capabilities | SIZE | MDTM  ( FluentFTP.FtpCapability )
>

Let me know if I did something wrong, since I don't see the full list of capabilities which you described.

Aside from this unexpected thing that I saw in the watch window, everything seems to be working now!

I'm still a bit surprised you are willing to go to this amount of effort for WuFtpd on HP-UX.  My workaround of overwriting the Capabilities during Connect() would have been pretty solid as well, and I don't expect to be using HP-UX very much longer anyway (if I can help it).

**Comment by robinrodricks** [comment]
Hi, could you check if the AssumeCapabilities method is taking effect? And
yes I corrected the implementation of fileExists do that is the thing
that's taking effect, still, I'd like to see my assumed capabilities show
up.

And no worries, I like solving easy issues. Most bugs are transient and
intermittent and nigh on impossible to solve. Yours isn't. Plus most bugs
are inter related and one fix leads to another :)

>

**Comment by dbeavon** [comment]
I'm kind of new to github and working with others on projects.  Are you asking me to clone the most recent copy of the github repository and step thru that method?  Or should I still be using the nuget binaries and put a breakpoint in the address-level debugger?

As I mentioned, I had used your prior binary/zip (Release.zip).  I put a breakpoint in FileExists, and detected the new capabilities that weren't there in the past (SIZE, MDTM).  When I say the capabilities weren't there in the past, I am referring to the version of the nuget package that we had previously settled on (19.2.2).  

Are you still implying that I should be seeing more capabilities than those two (SIZE, MDTM)?  If so I can help troubleshoot but probably not by using your binaries; I should start using your github source, right?

**Comment by robinrodricks** [comment]
Oh right. Nevermind. If you say those two capabilities appeared then the
new functionality is already at work. I'll release it officially on nuget
soon. I think we're done here. Thanks!

PS: you have an incredible amount of patience to debug a third party
library (this one) and to help me fix it too. Cheers!

**Comment by dbeavon** [comment]
Thanks for the help, and for the improved functionality of wuftpd.  

I will have to look at your code to see the rules by which certain features become available in different servers.  It seems like the best approach for this would be via FEAT, but in the absense of that I suppose you have to rely on some rules that are based on the name of the ftp server, os, or whatever.  Given that HP-UX is at the end of its life, you probably won't have to change those rules much going forward!  (Although I'm not sure how popular wuftpd is outside of HP-UX.)

Thanks again, let me know if you want me to push the close button.  Otherwise feel free to do that anytime.

**Comment by robinrodricks** [comment]
Released in [FluentFTP 23.1.0](https://www.nuget.org/packages/FluentFTP/23.1.0). Go grab it from nuget!
